[docs][pipes] Subprocess section (#17108)
dagster-io/dagster: docs/content/guides/dagster-pipes/subprocess/create-subprocess-asset.mdx

---
title: "Using Dagster Pipes, Part 1: Define a Dagster asset to invoke subprocess | Dagster Docs"
description: "Learn how to create a Dagster asset that invokes a subprocess that executes external code."
---

# Define a Dagster asset that invokes a subprocess

<Note>

This is part one of the [Using Dagster Pipes](/guides/dagster-pipes/subprocess) tutorial. If you are looking for how to modify your existing code that is already being orchestrated by Dagster, you can jump to <a href="/guides/dagster-pipes/subprocess/modify-external-code">Part 2: Modify external code</a>.

</Note>

> This part talks about how to create a Dagster asset that, in its execution function, opens a Dagster pipes session and invokes subprocess that executes some external code.

Author: we say "execution function" when referring the python function that's decorated by the @asset or the function that actually represents the execution of the asset. i've seen this in sensor and schedule docs when we talk about the schedule decorated function. but it's more for consistency within our docs but not a super technical accurate term

## Step 1: Define a Dagster asset that executes a subprocess

1. Create a new file named `dagster_subprocess_pipes.py` in the same directory as the `external_code.py` file you created earlier in the [Prerequisites](/guides/dagster-pipes/subprocess#prerequisites) step.
2. First, you’ll define the asset. Copy and paste the following into the file:

```python file=/guides/dagster/dagster_pipes/subprocess/dagster_subprocess_pipes.py startafter=start_asset_marker endbefore=end_asset_marker lines=-16
import shutil

from dagster import (
    AssetExecutionContext,
    MaterializeResult,
    PipesSubprocessClient,
    asset,
    file_relative_path,
)

@asset
def subprocess_asset(
    context: AssetExecutionContext, pipes_subprocess_client: PipesSubprocessClient
) -> MaterializeResult:
    cmd = [shutil.which("python"), file_relative_path(__file__, "external_code.py")]
```

Here’s what we did in this example:

- Created an asset named `subprocess_asset`

> - Provided <PyObject object="AssetExecutionContext" /> as the `context` argument to the asset. This object provides system information such as resources, config, and logging. We’ll come back to this a bit later in this section.

Author: we can do a pass all together later.

> - Provided <PyObject object="AssetExecutionContext" /> as the `context` argument to the asset. This object provides system information such as resources, config, and logging. We’ll come back to this a bit later in this section.

Reviewer: Sounds good!

---
title: "Using Dagster Pipes, Part 1: Define a Dagster asset that invokes subprocess | Dagster Docs"
description: "Learn how to create a Dagster asset that invokes a subprocess that executes external code."
---

# Using Dagster Pipes, part 1: Define a Dagster asset that invokes a subprocess

<Note>

This is part one of the [Using Dagster Pipes](/guides/dagster-pipes/subprocess) tutorial. If you are looking for how to modify your existing code that is already being orchestrated by Dagster, you can jump to <a href="/guides/dagster-pipes/subprocess/modify-external-code">Part 2: Modify external code</a>.

</Note>

In this part of the tutorial, you'll create a Dagster asset that, in its execution function, opens a Dagster pipes session and invokes a subprocess that executes some external code.

> ## Step 1: Define the Dagster asset

Reviewer: ```suggestion
---

## Step 1: Define the Dagster asset
```

```python file=/guides/dagster/dagster_pipes/subprocess/dagster_subprocess_pipes.py startafter=start_asset_marker endbefore=end_asset_marker lines=-16
import shutil

from dagster import (
    AssetExecutionContext,
    MaterializeResult,
    PipesSubprocessClient,
    asset,
    file_relative_path,
)

@asset
def subprocess_asset(
    context: AssetExecutionContext, pipes_subprocess_client: PipesSubprocessClient
) -> MaterializeResult:
    cmd = [shutil.which("python"), file_relative_path(__file__, "external_code.py")]
```

Here’s what we did in this example:

- Created an asset named `subprocess_asset`
- Provided <PyObject object="AssetExecutionContext" /> as the `context` argument to the asset. This object provides system information such as resources, config, and logging. We’ll come back to this a bit later in this section.
- Specified a resource for the asset to use, `PipesSubprocessClient`. We’ll also come back to this in a little bit.
- Declared a command list `cmd` to run the external script. In the list:
  - First, found the path to the Python executable on the system using `shutil.which("python")`.

>   - Then, provided the file path to the file that we want to execute. In this case, it’s the `extern al_code.py` file that you created earlier.

Reviewer: This is still an issue for me: `extern al_code.py` ?

import shutil

from dagster import (
    AssetExecutionContext,
    MaterializeResult,
    PipesSubprocessClient,
    asset,
    file_relative_path,
)

@asset
def subprocess_asset(
    context: AssetExecutionContext, pipes_subprocess_client: PipesSubprocessClient
) -> MaterializeResult:
    cmd = [shutil.which("python"), file_relative_path(__file__, "external_code.py")]
    return pipes_subprocess_client.run(
        command=cmd, context=context
    ).get_materialize_result()
```

Let’s take a look at what this code does:

- The `PipesSubprocessClient` resource used by the asset exposes a `run` method.
- When the asset is executed, this method will synchronously execute the subprocess in in a pipes session, and it will return a `PipesClientCompletedInvocation` object.
- This object contains a `get_materialize_result` method, which you can use to access the <PyObject object="MaterializeResult" /> event reported by the subprocess. We'll talk about how to report events from the subprocess in the next section.
- Lastly, return the result of the subprocess.

---

## Step 2: Define a Definitions object

> To make the asset and subprocess resource loadable and accessible by Dagster's tools, such as the CLI, UI, and Dagster Cloud, you’ll create a `Definitions` object which contains them.

Reviewer: ```suggestion
To make the asset and subprocess resource loadable and accessible by Dagster's tools, such as the CLI, UI, and Dagster Cloud, you’ll create a <PyObject object="Definitions" /> object that contains them.
```

)
```

At this point, `dagster_subprocess_pipes.py` should look like the following:

```python file=/guides/dagster/dagster_pipes/subprocess/dagster_subprocess_pipes_finished.py
import shutil

from dagster import (
    AssetCheckSpec,
    AssetExecutionContext,
    Definitions,
    MaterializeResult,
    PipesSubprocessClient,
    asset,
    file_relative_path,
)

@asset(
    check_specs=[AssetCheckSpec(name="no_empty_order_check", asset="my_asset")],
)
def subprocess_asset(
    context: AssetExecutionContext, pipes_subprocess_client: PipesSubprocessClient
) -> MaterializeResult:
    cmd = [shutil.which("python"), file_relative_path(__file__, "external_code.py")]
    return pipes_subprocess_client.run(
        command=cmd, context=context
    ).get_materialize_result()

defs = Definitions(
    assets=[subprocess_asset],
    resources={"pipes_subprocess_client": PipesSubprocessClient()},
)
```

---

## Step 3: Run the subprocess from the Dagster UI

In this step, you’ll execute the subprocess asset you created in earlier steps from the Dagster UI.

1. In a new command line session, run the following to start the UI:

   ```bash

>    dagster dev

Author: i think it might be fine - we don't call out a lot of experimental warnings so i'd prefer not to do it explicitly for pipes.

   ```

> 2. Navigate to <`localhost:3000`>, where you should see the UI:

Author: this seems to be consistent in the docs. i updated it to be`<http://localhost:3000>` for better consistency
